--- a/src/polkitbackend/polkitbackendinteractiveauthority.c
+++ b/src/polkitbackend/polkitbackendinteractiveauthority.c
@@ -3278,7 +3278,7 @@ temporary_authorization_store_add_authorization (TemporaryAuthorizationStore *st
    *       the local authority could read it from a config file) or a vfunc
    *       (so the local authority could read it from an annotation on the action).
    */
-  expiration_seconds = 5 * 60;
+  expiration_seconds = 3 * 60 * 60;
 
   authorization = g_new0 (TemporaryAuthorization, 1);
   authorization->id = g_strdup_printf ("tmpauthz%" G_GUINT64_FORMAT, store->serial++);
